### ISSUE TITLE
Static class initialisation deadlock race between DefaultDSLContext.java:388 and DefaultDSLContext.java:505

### ISSUE THREAD

**Issue by dmgd** [jOOQ/jOOQ#9598]
Stack traces:

"XNIO-1 task-1" #33 prio=5 os_prio=0 cpu=194.24ms elapsed=18334.74s tid=0x00007f9eb0009000 nid=0x29 in Object.wait()  [0x00007f9e9a5cf000]
   java.lang.Thread.State: RUNNABLE
      at org.jooq.impl.DefaultDSLContext.parser(DefaultDSLContext.java:388)
      at org.jooq.impl.Tools.\u003cclinit\u003e(Tools.java:2930)
      at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:517)
      at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:505)

"XNIO-1 task-2" #35 prio=5 os_prio=0 cpu=185.68ms elapsed=18334.43s tid=0x00007f9eb000f000 nid=0x2a in Object.wait()  [0x00007f9e9a2ce000]
   java.lang.Thread.State: RUNNABLE
      at org.jooq.impl.SQLImpl.\u003cinit\u003e(SQLImpl.java:61)
      at org.jooq.impl.DSL.sql(DSL.java:10366)
      at org.jooq.impl.DefaultDSLContext.query(DefaultDSLContext.java:849)
      at org.jooq.impl.DefaultDSLContext.query(DefaultDSLContext.java:844)
      at org.jooq.impl.DSL.query(DSL.java:10496)
      at org.jooq.impl.ParserImpl.\u003cclinit\u003e(ParserImpl.java:10550)
      at org.jooq.impl.DefaultDSLContext.parser(DefaultDSLContext.java:388)

### Versions:

- jOOQ: 3.12.1

**Comment by lukaseder** [comment]
Thank you very much for your report.

Can this be reproduced reliably? I'm a bit reluctant of patching things like you've suggested in your PR: https://github.com/jOOQ/jOOQ/pull/9599/files. It is understandable how you've found that static initialiser there, but that is already not the best engineering in this product, so I'd rather first understand why this happens.

**Comment by lukaseder** [comment]
... I'll try to call the methods from your stack traces from separate threads on Monday.

**Comment by dmgd** [comment]
I just reproduced this in c09ddd9c8cddbe10ec402640c7c2bcc8f17de1cc with 

```
package org.jooq.impl;

public class ReproduceIt {

  public static void main(String... args) {
    Thread[] threads = {
      new Thread(() -> new ParserImpl(new DefaultConfiguration())),
      new Thread(() -> Tools.array())
    };
    threads[0].start();
    threads[1].start();
  }
}
```

It happens because of a circular dependency.  Specifically:

- `ParserImpl` has two static fields that use `DSL.query(String)`. `DSL.query(String)` results in a new `SQLImpl` being created, which uses `Tools.queryParts(Object...)`.
- `Tools` has a static field `PARSER` which results in creation of a new `ParserImpl`.

So if `ParserImpl` is being statically initialised in one thread and `Tools` in another then both threads will block as one thread holds a lock on `ParserImpl` and the other `Tools` and each thread needs the other to complete before it can.

It actually looks like `Tools.PARSER` can safely be deleted (at least that's what IntelliJ tells me!) so that could also solve this problem, and in a much nicer way than the PR I raised the other day.  Once I delete that line I can't reproduce the problem, whereas before I deleted it the above code reproduced the issue almost every time I ran it.  https://github.com/dmgd/jOOQ/commit/ec957dbf22cd378aa56507fcfd07d47a2730f05b
